### ISSUE TITLE
CSS ::part and ::theme pseudo elements

### ISSUE THREAD

**Issue by fergald** [w3ctag/design-reviews#230]
こんにちはTAG！

I'm requesting a TAG review of:

  - Name: CSS ::part and ::theme pseudo elements
  - Specification URL: https://drafts.csswg.org/css-shadow-parts/
  - Explainer, Requirements Doc, or Example code: https://meowni.ca/posts/part-theme-explainer/
  - Tests: None yet
  - Primary contacts: tabatkins, fergald

こんにちはTAG！

I'm requesting a TAG review of:

  - Name: CSS ::part and ::theme pseudo elements
  - Specification URL: https://drafts.csswg.org/css-shadow-parts/
  - Explainer, Requirements Doc, or Example code: https://meowni.ca/posts/part-theme-explainer/
  - Tests: None yet
  - Primary contacts: tabatkins, fergald

You should also know that...

The spec is still in flux on some details
The syntax for exposing parts from sub-components (especially for bulk-exposing) will probably change.
There will probably be a separate theme= attribute for use with the ::theme() selector.

We'd prefer the TAG provide feedback as (please select one):

  - [ ] open issues in our Github repo for each point of feedback
  - [ ] open a single issue in our Github repo for the entire review
  - [X] leave review feedback as a comment in this issue and @-notify [github usernames]

You should also know that...

The spec is still in flux.
The syntax for exposing parts from sub-components (especially for bulk-exposing) will probably change.
There will probably be a separate theme= attribute for use with the ::theme() selector.

Partial design doc for Blink:
 https://docs.google.com/document/d/1eFekJxcNlkKHN2BwFwiwv6FoLrlRnuct_TGB2cKomW8/edit#

We'd prefer the TAG provide feedback as (please select one):

  - [ ] open issues in our Github repo for each point of feedback
  - [ ] open a single issue in our Github repo for the entire review
  - [X] leave review feedback as a comment in this issue and @-notify [github usernames]

**Comment by dbaron** [comment]
So my initial reactions to the [explainer / blog post](https://meowni.ca/posts/part-theme-explainer/) (without yet looking at any of the relevant specs):
* the `::part()` pseudo-element feels appropriate; the thing it does corresponds to its syntax and its name
* the `part`-forwarding microsyntax appears complex, and I wonder if it has (or should have) common characteristics with anything else
* I'm a little more queasy about the `::theme()` pseudo-element, since the thing it does feels more like a combinator than a pseudo-element.

I'll see how these reactions hold up after reading further...

(And yes, I was around when @tabatkins introduced these to the CSS WG, but that doesn't mean I remember them!)

**Comment by dbaron** [comment]
I don't think my reactions changed significantly by reading the [spec](https://drafts.csswg.org/css-shadow-parts/), although:
* the spec made the part-forwarding microsyntax feel simpler.  Though one thought was that perhaps forwarding a part from a nested component should have a slightly more separate syntax from declaring a part.  (This was a reaction to the spec's "ISSUE 4: Want a way to forward a name with the same name, without repeating the name (currently requires `foo => foo`).")  That is, if `forward-parts` was an attribute separate from `part` then `forward-parts="id1, id2 => id3"` would expose part `id1` from the inner component on the outer one, with the same name.
* the spec explains `::theme()` in a very different way from the explainer, and it's not yet clear to me how interesting this difference is.  The text is very formal, without a summary explanation, so I've yet to dig into the details (e.g., what a [mode](https://dom.spec.whatwg.org/#dom-shadowroot-mode) of [`closed`](https://dom.spec.whatwg.org/#dom-shadowrootmode-closed) means, which I don't think was covered in the explainer).

**Comment by dbaron** [comment]
To express my concern about `::theme()` being more like a combinator in a different way:  I'm worried that `::part()` and `::theme()` are too closely related to each other to have unrelated names.

**Comment by fergald** [comment]
@dbaron I also think a separate attribute for the partmap is a good idea. Reasons

- UX
  - part names behave much like class names but comma-separated. If part= was only for naming then we could remove the unexpected commas
  - this would solve the problem of how you shortcut writing part="a=>a". It's partmap="a"
- API
  - I guess we will want an API for programmatic manipulation, similar to classList. For parts we have a list and a map to deal with, which makes mungeing strings even less attractive. The natural API would keep naming separate from mapping e.g. partList() and partMap(). So separating the attributes would git nicely with that.

**Comment by tabatkins** [comment]
Yeah, I'm liking this idea of putting the forwarding into a different attribute.

**Comment by rniwa** [comment]
Can we also get rid of `=>`? So that it would be `id1, id2 id3` instead of `id1, id2 => id3`? This would make parsing a lot easier. We also need to have a IDL attribute which exposes this attribute so that manipulation in JS doesn't involve string manipulation.

**Comment by tabatkins** [comment]
Hm, we could, but I like the `=>` glyph for readability. ^_^  It's not clear in `part="id2 id3"` which part is the element's sub-part and which is the part name it's exposing; I think `id2 => id3` communicates clearly that `id2` is *becoming* `id3`.

> We also need to have a IDL attribute which exposes this attribute so that manipulation in JS doesn't involve string manipulation.

Definitely.  Seems like a pretty simple thing to represent as a `maplike<DOMString, DOMString>`.

**Comment by rniwa** [comment]
> Hm, we could, but I like the => glyph for readability. ^_^ It's not clear in part="id2 id3" which part is the element's sub-part and which is the part name it's exposing; I think id2 => id3 communicates clearly that id2 is becoming id3.

I understand but I would very strongly object to that syntax because it would make parsing in JS a lot harder and creates a lot more complicated parsing rules. We can call for consensus if you'd like.

**Comment by domenic** [comment]
I think it'd be good to actually spec a parser so we could see how complicated/hard to parse in JS it would be.

**Comment by tabatkins** [comment]
Yeah, the parsing is very easy, even if you're doing it by hand rather than with a regex.  (With regex the two situations are basically identical.)

**Comment by fergald** [comment]
I've filed a bunch of separate issues arising from this discussion and the F2F in Tokyo

* [2410 - separate part naming and part forwarding into 2 attributes](https://github.com/w3c/csswg-drafts/issues/2410)
* [2411 - decide on forwarding micro-syntax for partmap](https://github.com/w3c/csswg-drafts/issues/2411)
* [2412 - fully specify the parser for the forwarding micro-syntax](https://github.com/w3c/csswg-drafts/issues/2412)
* [2414 - IDL for attributes](https://github.com/w3c/csswg-drafts/issues/2414)

There was an existing issue for
* [2381 - Let ::theme() penetrate closed shadows](https://github.com/w3c/csswg-drafts/issues/2381)

**Comment by tabatkins** [comment]
Thanks, @fergald!

**Comment by dbaron** [comment]
I think this review is mostly done, but I think it's probably worth another quick pass once the edits mentioned above happen.  I'm curious if you have an expectation of when that's likely to be so that we can schedule this to come back to the TAG's agenda at an appropriate time?

I'm also still a little concerned about the naming, i.e., the issue that `::part()` and `::theme()` are so similar yet have unrelated names.  I'm curious if others also see this as an issue or are others unconcerned by it?

**Comment by fergald** [comment]
Thanks. I'll start working on the spec changes. Assuming things go smoothly, I'd hope to have them in next week some time.
